### PR TITLE
Add nix flake

### PR DESCRIPTION
**Pull request by goodlyrottenapple**
This PR adds preliminary support for building with nix, towards supporting installs via `kup` ( #147). To test, run:

```
nix shell .
```

inside the repo and `kavm` should build and be available on the PATH... stuff might still be broken when trying to run kavm tho...

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            ] ./.;
+
+            dontUseCmakeConfigure = true;
+
+            postPatch = ''
+              substituteInPlace ./Makefile \
+                --replace '$(MAKE) build -C $(PY_KAVM_DIR)' 'echo "Skip py-kavm"' \
+                --replace 'py-kavm $(VENV_DIR)/pyvenv.cfg' ' ' \
+                --replace 'plugin-deps $(hook_includes)' '$(hook_includes)'
+            '';
```
Comment (goodlyrottenapple) [inline flake.nix:108]: this is a bot hacky. basically I am removing any dependencies on `py-kavm` from the makefile as we build it separately using `poetry2nix` perhaps a refactor would be in order such that we can disable building of `py-kavm` via the Makefile through a flag. Same thing for `plugin-deps`, i.e. we build beforehand and don't want the Makefile to try to re-build again.

**Review by ehildenb — APPROVED** [review]
I approved to keep you unblocked, but one of @geo2a and @nwatson22 should confirm that this builds on their systems too! Ideally both!

**Comment by geo2a** [comment]
@goodlyrottenapple I can confirm that `nix build .` builds `kavm` and it becomes available on `PATH`. However, calling `kavm kompile ...` (as in the Makefile's `build` target) to build the semantics using the nix-build `kavm` does not work. It goes through to the point of linking the hook's shared objects in the llvm backend's `interpreter` and then fails.

**Comment by geo2a** [comment]
We had a call with @goodlyrottenapple and managed to figure out the necessary changes to make the set-up work. I think we can merge now!

File: flake.nix (hunk 1 above)
Comment (ehildenb) [reply flake.nix:112]: Well, I would like to move towards a future with less Makefile and more kbuild + Nix. So I think for now a hacky solution is OK.

**Pull request merged**